.github/workflows/ci.yml

```
@@ -263,6 +263,110 @@ jobs:
         run: |
           py -3 test\fuzz\test_runner.py --par %NUMBER_OF_PROCESSORS% --loglevel DEBUG %RUNNER_TEMP%\qa-assets\fuzz_corpora
 
+  windows-cross:
+    name: 'Linux->Windows cross, no tests'
+    runs-on: ubuntu-latest
+    if: ${{ vars.SKIP_BRANCH_PUSH != 'true' || github.event_name == 'pull_request' }}
+
+    env:
+      FILE_ENV: './ci/test/00_setup_env_win64.sh'
+      DANGER_CI_ON_HOST_FOLDERS: 1
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set CI directories
+        run: |
+          echo "CCACHE_DIR=${{ runner.temp }}/ccache_dir" >> "$GITHUB_ENV"
+          echo "BASE_ROOT_DIR=${{ runner.temp }}" >> "$GITHUB_ENV"
+          echo "DEPENDS_DIR=${{ runner.temp }}/depends" >> "$GITHUB_ENV"
+          echo "BASE_BUILD_DIR=${{ runner.temp }}/build" >> "$GITHUB_ENV"
+
+      - name: Depends cache
+        uses: actions/cache@v4
+        with:
+          path: ${{ env.DEPENDS_DIR }}/built
+          key: ${{ github.job }}-depends-${{ hashFiles('depends/**', 'ci/test/00_setup_env_win64.sh') }}
+
+      - name: Restore Ccache cache
+        id: ccache-cache
+        uses: actions/cache/restore@v4
+        with:
+          path: ${{ env.CCACHE_DIR }}
+          key: ${{ github.job }}-ccache-${{ github.run_id }}
+          restore-keys: ${{ github.job }}-ccache-
+
+      - name: CI script
+        run: ./ci/test_run_all.sh
+
+      - name: Save Ccache cache
+        uses: actions/cache/save@v4
+        if: github.event_name != 'pull_request' && steps.ccache-cache.outputs.cache-hit != 'true'
+        with:
+          path: ${{ env.CCACHE_DIR }}
+          key: ${{ github.job }}-ccache-${{ github.run_id }}
+
+      - name: Upload built executables
+        uses: actions/upload-artifact@v4
+        with:
+          name: x86_64-w64-mingw32-executables-${{ github.run_id }}
+          path: |
+            ${{ env.BASE_BUILD_DIR }}/src/**/*.exe
+            ${{ env.BASE_BUILD_DIR }}/test/config.ini
+
+  windows-native-test:
+    name: 'Windows, test cross-built'
+    runs-on: windows-2022
+    needs: windows-cross
+
+    env:
+      PYTHONUTF8: 1
+      TEST_RUNNER_TIMEOUT_FACTOR: 40
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Download built executables
+        uses: actions/download-artifact@v4
+        with:
+          name: x86_64-w64-mingw32-executables-${{ github.run_id }}
+
+      - name: Run bitcoind.exe
+        run: .\src\bitcoind.exe -version
+
+      - name: Run unit tests
+        run: |
+          .\src\test\test_bitcoin.exe -l test_suite
+          .\src\secp256k1\bin\exhaustive_tests.exe
+          .\src\secp256k1\bin\noverify_tests.exe
+          .\src\secp256k1\bin\tests.exe
+          .\src\univalue\object.exe
+          .\src\univalue\unitester.exe
```

https://github.com/bitcoin/bitcoin/pull/31176#issuecomment-2699970750 Seems that probably only the exit code of the last command matters here, one fix is to append ``&& ` `` to each command except the last:

```suggestion
      - name: Run unit tests
        run: |
          .\src\test\test_bitcoin.exe -l test_suite && `
          .\src\secp256k1\bin\exhaustive_tests.exe && `
          .\src\secp256k1\bin\noverify_tests.exe && `
          .\src\secp256k1\bin\tests.exe && `
          .\src\univalue\object.exe && `
```

---

Thanks! Taken.

In my initial implementation, I relied on the [fail-fast behavior](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference) in GitHub-hosted runners. However, it turns out that `$ErrorActionPreference = 'stop'` only applies to PowerShell cmdlets.

---

> I don't understand this. Don't the docs claim that `set -e` is set for bash/sh?

But it is not `bash` or `sh` in the Windows environment.

---

I think since we don't set `shell` for this specific step, we use [PowerShell by default on Windows](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell). Slightly unclear [documentation](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference) on what the expected exit code behavior is.

---

There might be some limited value in using cmd/PowerShell on Windows in order to execute in a closer-to-end-user environment and tolerate the "&& \`". But I'm not totally against switching to bash/scripts either.

---

Thanks!

I've added a commit, which sets Bash as the default shell globally.

---

Thread: https://github.com/bitcoin/bitcoin/pull/31176#discussion_r2000738410:

Nice! Hopefully that satisfies others in the thread.

Taking one step back, would it make sense to add a comment about why we call the test executables directly here instead of going through ctest like other jobs? I'm guessing we don't want to depend on ctest as we just download executables via artifacts.

---

Makes sense, certainly not blocking.
```cmake
# Can't use ctest here like other jobs as we don't have a CMake build tree.
```